### ISSUE TITLE
VWLineSimplifier Optimization

### ISSUE THREAD

**Issue by airbreather** [NetTopologySuite/NetTopologySuite#33]
Not exactly sure where I should be putting this, as I don't think it's worthy of a pull request just yet, but I've got a heavily optimized version of VWLineSimplifier that I'm tentatively calling "FastVWLineSimplifier".  The TODO caught my eye and I couldn't resist.

airbreather/NetTopologySuite@ee14fc978516541727de39c701389bc59f4130f5

That commit includes a test case that produces output similar to this on my machine:
Time (slow): 11.9271277033042 seconds
Time (fast): 0.0595896480771026 seconds

Notably, FastVWLineSimplifier appears to return perfectly identical results to what VWLineSimplifier would produce (doing a full replacement passed all the NUnit tests), though the implementation is obviously completely different.

I think all it should really need is some surface-level cleanup and general direction from you guys about what to do about it.

Thoughts?

**Comment by FObermaier** [comment]
Great! As we try to keep NTS in line with JTS I think we should inform @dr-jts about your improvement.

**Comment by DGuidi** [comment]
wow, thanks :)
to me, we should already integrate code into NTS, maybe in an "experimental" namespace, and then, if and when @dr-jts integrates code into JTS, adapt NTS code to the JTS one.

**Comment by DGuidi** [comment]
@airbreather feel free to open a pull request when you think code is stable enough

**Comment by dr-jts** [comment]
Nice...  

JTS already has a PriorityQueue:

http://sourceforge.net/p/jts-topo-suite/code/HEAD/tree/trunk/jts/java/src/com/vividsolutions/jts/util/PriorityQueue.java

Is that in NTS, and if so would it work for this?

**Comment by airbreather** [comment]
@dr-jts: It [appears so](https://github.com/NetTopologySuite/NetTopologySuite/blob/master/NetTopologySuite/Utilities/PriorityQueue.cs); the main thing I wanted to solve before making a proper pull request is to find the priority queue implementation that made the most sense to use.  This may be it.

**Comment by FObermaier** [comment]
@airbreather, would you try to adopt your code to using the PriorityQueue in the NetTopologySuite.Utilities namespace?

**Comment by airbreather** [comment]
@FObermaier yeah, working on it now.

**Comment by airbreather** [comment]
The issue I've been running into is that the algorithm needs much more than either the "Original" or the "Alternative" implementation in PriorityQueue.cs provides: it needs to be able to update the priorities of the nodes that are already in the queue and have the queue deal with it.  I don't really see a way around that, because whenever we delete a point, we always need to update the areas of the 1 or two points around it, unless we're deleting the last point we're allowed to delete.

So the solution is looking like it's going to need at least some modifications to PriorityQueue<T>, at the minimum to make it so that subclasses have enough access to do the manipulation that they need to do to make this work.

OrderedBag<T> from Wintellect.PowerCollections has *similar* semantics to PriorityQueue<T>, but it also has the same problem of not having a way for us to tell it to update the priority of things already in the queue.

Long story short, its not nearly as simple as "switch from this priority queue implementation to one that we already have" makes it sound.

**Comment by dr-jts** [comment]
Yeah, I was afraid of that... I think that's why I stopped short of using
PriorityQueue when I implemented VWS in the first place.

The right answer is to enhance PriorityQueue to support remove(), I
think.(I believe that's all that is strictly necessary, since nodes can
then be removed, have their priority updated, and then re-inserted?).  That
should be easy enough...

**Comment by airbreather** [comment]
That seems like a reasonable approach... slower than what I have here, but still O(log n) per deletion, and doesn't pollute the repository with otherwise-unnecessary external code.  I'll try that when I get home, should be able to have it by the end of Wednesday (UTC-5) one way or another.

**Comment by airbreather** [comment]
OK, airbreather/NetTopologySuite@3861c66 is my first take on it.  I don't see a way to keep it at O(n log n) without having something like PriorityQueueNode<T>.

Can anyone think of a reason I shouldn't go ahead with something like this?

**Comment by FObermaier** [comment]
From what I understand reading the latest posts I gather that extending the existing PriorityQueue will not be as fast as your initial choice.

So I'm wondering if it wouldn't make more sense to replace the existing PriorityQueue with your initial version and have it implement/mimic the original public interface.

**Comment by dr-jts** [comment]
Could you keep an internal hashmap containing the indexes of the items?
That way the same PriorityQueue interface can be maintained.

**Comment by DGuidi** [comment]
maybe @airbreather you can refactor current priorityqueue implementation in a way to abstract a common interface to apply to current code (with few changes) and to your implementation.
these changes then can be applied to JTS code as well.

**Comment by dr-jts** [comment]
Or would it be efficient to not remove modified CoordinatesWithArea, but
mark them as dead and insert new one with the new area?  Dead coordinates
would be discarded if encountered while pulling from the Queue.

**Comment by airbreather** [comment]
The way I see it, there are several factors to consider.  The relative importance really determines what approach is best:

1. How much we really care about external code that uses PriorityQueue<T> (it's public, after all).
2. Impact on the internal NTS consumers of PriorityQueue<T>.  Currently, this is just STRtree-related types, but consider that STRtree is itself used in performance-sensitive cases.
3. [com.vividsolutions.jts.util.PriorityQueue](http://tsusiatsoftware.net/jts/javadoc/com/vividsolutions/jts/util/PriorityQueue.html) seems to be heavily modeled off of [java.util.PriorityQueue](http://docs.oracle.com/javase/7/docs/api/java/util/PriorityQueue.html), which does have a [remove](http://docs.oracle.com/javase/7/docs/api/java/util/PriorityQueue.html#remove(java.lang.Object)), but which is [implemented O(N)](http://hg.openjdk.java.net/icedtea/jdk8/jdk/file/aa2c9ce06632/src/share/classes/java/util/PriorityQueue.java#l359) making it unusable in this instance (I am almost certain that the existing code will be faster than [this](https://github.com/airbreather/NetTopologySuite/commit/9d81ae5bbce50098e57e9e7757f1e9a075eb633f) in all instances if remove() were O(N))... it looks to me like JTS could easily switch to using java.util.PriorityQueue any time @dr-jts feels like it, whereas without a .NET equivalent, NTS would have to mimic it using something like [OrderedBag from PowerCollections](http://powercollections.codeplex.com/SourceControl/latest#Source/PowerCollections/OrderedBag.cs), which wouldn't be a fantastic fit here because it's implemented using a Red-Black tree.
4. "Yuckiness" factor of having two heap-based PriorityQueue types in NTS (/JTS).
5. Performance.

The best safe solution I can think of without knowing how important any of those is ("safe" assuming that BlueRaja's implementation is correct in all the cases we care about) is to bring in something similar to BlueRaja's implementation and keep PriorityQueue<T> around, but (optionally) replace the existing PriorityQueue<T>'s implementation with one that wraps an instance of the other class.  VWLineSimplifier would use the fancy features offered by the new queue type rather than a PriorityQueue<T>.

Here's how that holds up by all the above checklist items:

1. External code will still work as-is; if we replace the existing PriorityQueue<T> implementation with a wrapper around the other one, then it might be very slightly slower and see an increased number of heap allocations.
2. NTS code is affected the same way external code would be.
3. @dr-jts can choose to port all, some, or none of the VWLineSimplifier enhancement (which includes the new queue type) as a decision that is completely separate from the decision of whether or not to drop com.vividsolutions.jts.util.PriorityQueue in favor of java.util.PriorityQueue.
4. "Yuckiness" is increased unless the "optional" replacement is done.
5. VWLineSimplifier is as fast as I can get it without cheating; as mentioned in first two checklist items, though, doing the "optional" replacement might have (relatively small) negative performance implications for STRtree.

OK, based on all the above, I'm going to leave this open for a few days, but otherwise work on turning this into a more professional level of code suitable for a pull request.

Specific responses to specific comments since my last one will come in my next comment.

**Comment by airbreather** [comment]
In order:

@FObermaier, yes and no.  A heap-based priority queue is a heap-based priority queue.  This data structure is not (necessarily) optimized for efficient operations that aren't at the head of the queue.  But with a "pointer" (in this case, the index) to the item to remove / priority-update, we can leverage the same algorithms used to find the priority of an item that's being newly inserted in O(log n) time.

The "problem" arises when we try to apply that knowledge to PriorityQueue&lt;T&gt;: if all we have is a reference to the data being stored in the queue and not a pointer to it, we have to do a linear search to find exactly where it's stored every time we want to do an update.  I wouldn't be exaggerating to claim that the linear search would more than erase all performance gains from doing it this way.

So yes, I'm going with something very similar to what you've suggested.

@dr-jts: technically yes.  Not only would this be notably slower and require O(N) additional memory, but it would also (unintuitively?) require changing NetTopologySuite.Index.Strtree.BoundablePair&lt;TItem&gt; to override Equals(object) and GetHashCode() to be consistent with the results of IComparable&lt;BoundablePair&lt;TItem&gt;&gt;.  This kind of thing would also apply to (hypothetical) external users of PriorityQueue&lt;T&gt;.  This really scares me, because the problems that would arise if they didn't do so would likely be extremely subtle, and it might even be difficult to detect that a problem exists at all depending on the use case.

@DGuidi: yes, this is something similar to what I'm going to be doing.

@dr-jts: again, I think you're technically correct, "delay-deletion" is really tempting in a vacuum.  This time, I'm worried about general applicability and memory usage.

w.r.t. general applicability: it feels like the "delay-deletion" solution could make it possible that a given sequence of operations causes the memory consumption of the queue to grow without bound.  While that shouldn't be the case for VWLineSimplifier, it's not hard to imagine someone using PriorityQueue&lt;T&gt; for something that can't guarantee that.  So unless someone can convincingly argue that this is impossible, doing that would basically make this queue only viable for this problem.

w.r.t. memory usage: even if it were restricted to VWLineSimplifier so there's no chance for "grow without bound", I can still see memory usage growing immensely, and really annoying is the fact that this gets **more** likely to happen the more useful the results of this would be: the memory complexity doesn't change from O(N), but the scalar factor almost triples; considering that one of the big use cases I can see for simplification is to *reduce* memory consumption overall, I find something with those properties to be a really hard sell when there's an alternative that's admittedly a bit more annoying to us initially, but then it becomes not-annoying afterward.

**Comment by airbreather** [comment]
Going to close this now that there's a pull request.
